FPU fixes for the simulator

bartoldeman opened the pull request:
This is a collection of various fixes to fp87-sim.c. It's not perfect yet but getting there.
There are a few more standalone tests from QEMU that were helpful (compiled with DJGPP out of the box), in
https://gitlab.com/qemu-project/qemu/-/tree/master/tests/tcg/i386
@andrewbird you may find them useful, haven't gone through all of them yet though (see commit msgs for which ones I used)

bartoldeman committed:
fp87-sim: test for pseudo-normals/denormals

Those are esoteric values, explained here:
https://developers.redhat.com/blog/2021/05/12/mostly-harmless-an-account-of-pseudo-normal-floating-point-numbers
and don't fall in the usual fpclassify categories.

This change fixes qemu's test-i386-fxam.c test.

bartoldeman committed:
fp87-sim: use long double floating point constants

M_PI and friends used double precision which is not sufficient for long double.
This fixes the mcpdiag constant tests.

bartoldeman committed:
fp87-sim: handle some NAN and infinity cases for fscale

Fixes the QEMU test-i386-fscale.c test

bartoldeman committed:
fp87-sim: detect empty floating point slot in FXAM

bartoldeman committed:
fp87-sim: set rounding on entry and when control word changes.

Fixes more tests in mcpdiag.exe, and qemu's test-i386-fldcst.c.

stsp commented:
> It's not perfect yet but getting there.

You may want to use draft, and
remove it when need be. :)

bartoldeman committed:
fp87-sim: fix fbstp for special/out-of-range input values

Fixes QEMU's test-i386-fbstp.c test.

bartoldeman committed:
fp87-sim: don't mask out just set inexact exception flag!

This fixes many mcpdiag failures and some QEMU test-i386-fp-exceptions.c
tests.

andrewbird commented:
@bartoldeman thanks for the link to the fpu tests, I'll try to add them in the next few days.

bartoldeman committed:
fp87-sim: fix fprem/fprem1 simulation

Using a combination of fmodl(), remainderl(), and remquotl() it now passes the
QEMU test case test-i386-fprem.c, except for one issue with stack underflow.

bartoldeman committed:
fp87-sim: use rintl() instead of round_double()

Now we use fesetround() we can just use rintl() all the time, it does the
right thing.

bartoldeman committed:
fp87-sim: reorganize masked exception handling

We now keep the masked exception state on the native FPU except if
access to the status word is needed (FSTSW/FSTENV/FSAVE)
and if some emulator code can cause exceptions, in which case
we need to save before and clear after.

This fixes all remaining mcpdiag failures (except one, but that also fails
on real hardware), and QEMU's test-i386-fp-exceptions.c test.

bartoldeman committed:
fp87-sim: use symbolic constants for FPU status word

bartoldeman committed:
fp87-sim: fyl2xp1/fyl2x/f2xm1 log/exp fixes.

Fixes many of the QEMU tests for those instructions, with remaining ones
only differing by a few ULP.

bartoldeman commented:
It now passes all of mcpdiag, except for one test that also fails on real hardware (because Intel changed their fsin/fcos algorithms a little some years ago).

The extra QEMU tests have some failures but it's either minor (tiny errors) or fairly obscure (stack underflow / denormal bits of FPU status word), so this is good to go, further enhancements can always go in a new PR.

stsp commented:
Thanks!
That's really cool.
